### PR TITLE
Problem: interchain-accounts module is not integrated

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #540

Solution:
- integrate ica controller module
- add integration test, refactor the ibc integration test using builtin capabilities of pystarport

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

**Review by tomtau — COMMENTED** [review]
x/icaauth seems to be for the "host" chain , so perhaps that's not needed in the context of https://github.com/crypto-org-chain/cronos/issues/539 ? i.e. x/icaauth would need some patches in Ethermint to support eth accounts and transactions?

**Review by tomtau — COMMENTED** [review]
if icaauth is really needed even for controller, could it be imported from crypto.org chain repo (if duplicate / identical)?

**Comment by devashishdxt** [comment]
> if icaauth is really needed even for controller, could it be imported from crypto.org chain repo (if duplicate / identical)?

We may need custom hooks in cronos when receiving acknowledgement.

**Comment by yihuang** [comment]
> if icaauth is really needed even for controller, could it be imported from crypto.org chain repo (if duplicate / identical)?

there are several minor differences right now:
- chain-main master branch has moved to ibc-go master for the relayer incentivization support, there are api breaking changes.
- cronos's version removed the unused `storeKey`.
- namespace change in the rest api, `/cronos/icaauth/v1/interchain_account_address` vs `/chain-main/icaauth/v1/interchain_account_address`.

**Comment by yihuang** [comment]
> > if icaauth is really needed even for controller, could it be imported from crypto.org chain repo (if duplicate / identical)?
> 
> We may need custom hooks in cronos when receiving acknowledgement.

yeah, we might need that to integrate with smart contract.

File: integration_tests/pytest.ini [hunk 1]
```diff
@@ -1,3 +1,2 @@
 [pytest]
-asyncio_mode = strict
```
Comment (yihuang) [inline integration_tests/pytest.ini:2]: remove a warning.

File: integration_tests/test_ibc.py [hunk 2]
```diff
@@ -1,57 +1,85 @@
...
+    "start-cronos"
+    path = tmp_path_factory.mktemp("ibc")
+    gen = setup_custom_cronos(path, 26700, Path(__file__).parent / "configs/ibc.yaml")
+    cronos = next(gen)
+    try:
+        chainmain = Chainmain(cronos.base_dir.parent / "chainmain-1")
+        hermes = Hermes(cronos.base_dir.parent / "relayer.toml")
+        # wait for grpc ready
+        wait_for_port(ports.grpc_port(chainmain.base_port(0)))  # chainmain grpc
+        wait_for_port(ports.grpc_port(cronos.base_port(0)))  # chainmain grpc
```
Comment (yihuang) [inline integration_tests/test_ibc.py:41]:
```suggestion
        wait_for_port(ports.grpc_port(cronos.base_port(0)))  # cronos grpc
```

**Comment by yihuang** [comment]
> if icaauth is really needed even for controller, could it be imported from crypto.org chain repo (if duplicate / identical)?

I'll try to simplify it, I guess it's possible to move some functionalities to cronos module itself.

**Comment by yihuang** [comment]
> > if icaauth is really needed even for controller, could it be imported from crypto.org chain repo (if duplicate / identical)?
> 
> I'll try to simplify it, I think it's possible to merge the functionalities to cronos module itself.

ah, I think all the codes in icaauth module are actually for the controller side, even if merge the modules, it can't be simplified:
- it provides the tx msgs/queries/clis for the controller size.
- it provides the ibc module to handle capability and permissions for the controller side.

it's actually not very useful for host side I think, it's critical for the controller side.

**Comment by yihuang** [comment]
Maybe the name `icaauth` is confusing here, to my understanding, it's something to complete the ica controller, the controller module provided in `ibc-go` only provides the ibc middleware and keeper, and the `icaauth` provides the tx msgs, query handlers, cmds, ibc module.

**Comment by tomtau** [comment]
maybe `ica`, given it's the shortest? :)

**Comment by yihuang** [comment]
> maybe `ica`, given it's the shortest? :)

how about `icactl`? it emphasis it's the controller side logic, and don't directly conflict with the name `icacontroller`.

**Comment by tomtau** [comment]
`icacontroller` is fine too.
@devashishdxt do you have any preference (in the case you plan to refactor the name in chain-main to make it aligned)?

**Comment by devashishdxt** [comment]
>  `icacontroller` is fine too.
> 
> @devashishdxt do you have any preference (in the case you plan to refactor the name in chain-main to make it aligned)?

I don't think we can use `icacontroller` because it conflicts with the module in `ibc-go`. `icactl` sounds good.

**Review by tomtau — COMMENTED** [review]
I can have another round of review next week unless it's urgent

**Comment by yihuang** [comment]
> I can give another round of review next week unless it's urgent

I think not urgent

File: app/app.go [hunk 3]
```diff
@@ -301,37 +318,24 @@ func New(
...
-			evidencetypes.StoreKey, capabilitytypes.StoreKey,
-			feegrant.StoreKey, authzkeeper.StoreKey,
-			// ibc keys
-			ibchost.StoreKey, ibctransfertypes.StoreKey,
-			// ethermint keys
-			evmtypes.StoreKey, feemarkettypes.StoreKey,
-			// this line is used by starport scaffolding # stargate/app/storeKey
-			cronostypes.StoreKey,
-		)
+		storeKeys = append(storeKeys, gravitytypes.StoreKey)
```
Comment (tomtau) [inline app/app.go:336]: the order of store keys doesn't matter I guess? (before gravity was before cronos)

File: app/app.go [hunk 4]
```diff
@@ -708,6 +735,25 @@ func New(
...
+		panic(fmt.Sprintf("failed to read upgrade info from disk %s", err))
+	}
+
+	if upgradeInfo.Name == planName && !app.UpgradeKeeper.IsSkipHeight(upgradeInfo.Height) {
+		storeUpgrades := storetypes.StoreUpgrades{
+			Added: []string{icacontrollertypes.StoreKey},
+		}
+
+		// configure store loader that checks if version == upgradeHeight and applies store upgrades
+		app.SetStoreLoader(upgradetypes.UpgradeStoreLoader(upgradeInfo.Height, &storeUpgrades))
```
Comment (tomtau) [inline app/app.go:754]: should this upgrade handle have the gravity bridge stuff as well?

File: nix/sources.json [hunk 5]
```diff
@@ -1,4 +1,16 @@
 {
+    "chain-main": {
```
Comment (tomtau) [inline nix/sources.json:2]: maybe this can use the Nix flake instead?

File: x/icactl/keeper/keeper.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	if err != nil {
+		return err
+	}
+
+	packetData := icatypes.InterchainAccountPacketData{
+		Type: icatypes.EXECUTE_TX,
+		Data: data,
+	}
+
+	timeoutTimestamp := ctx.BlockTime().Add(timeoutDuration).UnixNano()
```
Comment (tomtau) [inline x/icactl/keeper/keeper.go:80]: is the arbitrary user-set timeout desirable? I recall there was one issue that for ICA, the channel will be closed after the timeout

Comment (tomtau) [reply x/icactl/keeper/keeper.go:80]: oh, this usually should go via the other functions that has `k.MinTimeoutDuration` ... maybe good to highlight that

File: app/app.go [hunk 7]
```diff
@@ -708,6 +735,25 @@ func New(
 
 	app.SetEndBlocker(app.EndBlocker)
 
+	planName := "v0.8.0"
+	app.UpgradeKeeper.SetUpgradeHandler(planName, func(ctx sdk.Context, _ upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+		return app.mm.RunMigrations(ctx, app.configurator, fromVM)
```
Comment (tomtau) [inline app/app.go:740]: should it set the minimum timeout duration?

File: x/icactl/module_ibc.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+// OnRecvPacket implements the IBCModule interface
+func (am IBCModule) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) ibcexported.Acknowledgement {
+	// https://github.com/cosmos/ibc-go/blob/v3.0.0/docs/apps/interchain-accounts/auth-modules.md#ibcmodule-implementation
+	return channeltypes.NewErrorAcknowledgement("cannot receive packet on controller chain")
+}
+
```
Comment (tomtau) [inline x/icactl/module_ibc.go:114]: I guess some of these functions should then be leveraged by x/cronos for EVM integrations (to call some EVM functions)?

**Review by tomtau — APPROVED** [review]
I haven't read through the generated pb files, but others look ok

File: app/app.go (hunk 4 above)
Comment (yihuang) [reply app/app.go:754]: Is it decided that if gravity-bridge is to release in `v0.8.0`?

Comment (tomtau) [reply app/app.go:754]: I don't follow the release plan of Cronos, so I'm not sure

File: nix/sources.json (hunk 5 above)
Comment (yihuang) [reply nix/sources.json:2]: Right now, the integration test shell in still managed with traditional nix, the last time I tried to merge all the stuff into the default flake, it'll build unnecessary stuff when building the binary itself.

File: app/app.go (hunk 3 above)
Comment (yihuang) [reply app/app.go:336]: yes, the `NewKVStoreKeys` returns a map.

File: x/icactl/keeper/keeper.go (hunk 6 above)
Comment (yihuang) [reply x/icactl/keeper/keeper.go:81]: sure, added a comment.

File: x/icactl/module_ibc.go (hunk 8 above)
Comment (yihuang) [reply x/icactl/module_ibc.go:114]: the `icactl` is only part of the controller stack, for the host side, we need to add another middleware when nesserary.

File: app/app.go (hunk 4 above)
Comment (tomtau) [reply app/app.go:754]: for ref from the call today: `v0.8.0` (or `v1.0.0`? 😁) is planned for September... if GB can be finalized before that, it's to be included

File: app/app.go (hunk 7 above)
Comment (tomtau) [reply app/app.go:740]: if it's from that chain-main value, I guess it should be ok... it was based on a discussion @calvinaco @devashishdxt decided referencing the relayer time backlog from the past

File: x/icactl/module_ibc.go (hunk 8 above)
Comment (tomtau) [reply x/icactl/module_ibc.go:114]: I meant in the controller context... if a smart contract is sending an IBC message that controls an account on another chain, but it times out... how is/will the contract be notified of it?

File: nix/sources.json (hunk 5 above)
Comment (yihuang) [reply nix/sources.json:2]: done, added the integration test.

File: app/app.go (hunk 7 above)
Comment (yihuang) [reply app/app.go:740]: > the default value is 1hour, looks good? I'm reenabling the upgrade test case to assert this.

done, added the upgrade integration test.

File: app/app.go (hunk 4 above)
Comment (yihuang) [reply app/app.go:754]: I guess we can complete the upgrade handler later when release plan is more finalized.

**Pull request merged**